sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSparkPath.scala

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import java.net.URI
+
+import org.apache.hadoop.fs.{FileStatus, Path}
+
+/**
+ * This is copied from the `SparkPath` in Apache Spark.
+ */
+final case class GpuSparkPath private (private val underlying: String) {
```

Comment:
I'm not a fan of this class and how it's used.  It's always used as a throwaway object to get to something else.  The whole point of this class should be "I need to track something as a path, not just a string", but it's never used that way.  No data structures were updated to use GpuSparkPath instead of strings, all the instances are ephemeral.

Therefore given the way this is going, I'd much rather ditch this class entirely and go with the following approach.  Please feel free to poke holes in it:

- Everywhere we need to deal with something that might be a SparkPath, tack on `.toString` so we know it will always be a string.  Common plugin code always deals with paths as strings.
- Add a shim static function named `toSparkPath` that takes a string and either returns a String on Spark versions that do not have SparkPath or returns SparkPath on Spark versions that do.

Those to methods should give us the ability to read and wield items that may be SparkPath's and create SparkPath instances when necessary.

Reply:
> The whole point of this class should be "I need to track something as a path, not just a string", but it's never used that way.

Thank you for the review. Actually, the `SparkPath` doesn't do any check when constructing from `String` (https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/paths/SparkPath.scala), which confuses me either.

The rationale of https://github.com/apache/spark/commit/faedcd91 is that  (from the PR description):
```
Given `val str = "s3://bucket/path with space/a"` There is a difference between `new Path(str)` and `new Path(new URI(str))`, and thus a difference between `new URI(str)` and `new Path(str).toUri`.
Both `URI` and `Path` are symmetric in construction and `toString`, but are not interchangeable. Spark confuses these two paths (uri-encoded vs not). This PR attempts to use types to disambiguate them
```

So that the author introduced `SparkPath` and two functions `SparkPath.toPath` and  `SparkPath.toUri` to try to get rid of the confusing thing. 

>Everywhere we need to deal with something that might be a SparkPath, tack on .toString so we know it will always be a string. Common plugin code always deals with paths as strings.
Add a shim static function named toSparkPath that takes a string and either returns a String on Spark versions that do not have SparkPath or returns SparkPath on Spark versions that do.

Both of the 2 ideas are good and practical. However, the point of this audit is to avoid creating a `Path` by calling the constructor explicitly, but using `SparkPath.toPath` instead. I guess we should follow the change in Spark.